### ISSUE TITLE
Dictionaries can only hold a maximum of 4096 indices

### ISSUE THREAD

**Issue by toddfarmer** [toddfarmer/arrow-migration#1361]
***Note**: This issue was originally created as [ARROW-1407](https://issues.apache.org/jira/browse/ARROW-1407). Please see the [migration documentation](https://gist.github.com/toddfarmer/12aa88361532d21902818a6044fda4c3) for further details.*

### Original Issue Description:
Dictionaries seem to only be able to hold 4096 indices, meaning only vectors with 4096 values or less can be turned into dictionaries. The image attached is a stack trace of what happens when try to encode a dictionary with a vector containing 4097 strings, and a dictionary containing two distinct values. 

Basically the error can be traced to line 95 of DictionaryEncoder.java (`setter.invoke(mutator, i, encoded);`). It seems that the indices array which hold the encoded values is allocated on line 84 as `indices.allocateNew()` and it seems that `allocateNew()` only allocates 4096 bytes of data initially. The code runs if there are 4096 rows of data or less. Anymore and the same error is given.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1407?focusedCommentId=16138958) by Emilio Lahr-Vivaz (elahrvivaz):*
FYI, you don't have to use the DictionaryEncoder class. If you don't mind mapping your dictionary values yourself, you can do something like:

```java
NullableIntVector vector = new FieldType(true, MinorType.INT.getType, dictionaryEncoding).createNewSingleVector(name, allocator, callBack);
vector.getMutator().setSafe(i, j);
```

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1407?focusedCommentId=16149459) by Wes McKinney (wesm):*
`[~icexelloss]` or `[~alphalfalfa]` any interest in taking a look at this?

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1407?focusedCommentId=16149498) by Li Jin (icexelloss):*
I'll take this one.

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1407?focusedCommentId=16150555) by Li Jin (icexelloss):*
PR: https://github.com/apache/arrow/pull/1024

**Comment by toddfarmer** [comment]
***Note**: [Comment](https://issues.apache.org/jira/browse/ARROW-1407?focusedCommentId=16151126) by Wes McKinney (wesm):*
Issue resolved by pull request 1024
<https://github.com/apache/arrow/pull/1024>
